src/test/java/com/styra/opa/OPATest.java

```diff
@@ -220,6 +230,106 @@ public void testOPAEcho() {
         assertEquals(expect, result);
     }
 
+    @Test
+    public void testOPAEchoWithCustomClient() {
+        // Normally, OPAClient handles instantiating the OPAHTTPClient under
+        // the hood to accommodate custom headers. This test ensures that
+        // passing it to the constructor directly works, since if the client
+        // isn't adding the headers properly then OPA's auth will bounce
+        // the requests.
+
+        HTTPClient httpc = new OPAHTTPClient(headers);
+        OPAClient opa = new OPAClient(address, httpc);
+        Map result = Map.ofEntries(entry("unit", "test"));
+        Map expect = Map.ofEntries(entry("hello", "world"), entry("foo", Map.ofEntries(entry("bar", testIntegerA))));
+
+        try {
+            result = opa.evaluate("policy/echo", Map.ofEntries(
+                entry("hello", "world"),
+                entry("foo", Map.ofEntries(entry("bar", testIntegerA)))
+            ));
+        } catch (OPAException e) {
+            System.out.println("exception: " + e);
+            assertNull(e);
+        }
+
+        assertEquals(expect, result);
+    }
+
+    @Test
+    public void testLatencyMeasurement() {
+        Logger logger = Logger.getLogger("testLatencyMeasurementLogger");
+        RecordingHandler handlerObj = new RecordingHandler();
+        handlerObj.setLevel(ALL);
+        logger.addHandler(handlerObj);
+        logger.setLevel(ALL);
+        logger.setUseParentHandlers(false);
+
+        OPALatencyMeasuringHTTPClient httpc = new OPALatencyMeasuringHTTPClient(logger, headers);
+        httpc.setLatencyMeasurementFormat("LATENCY MEASUREMENT #{0,number,#}#{1}#");
+        httpc.setLatencyMeasurementLogLevel(INFO);
+        OPAClient opa = new OPAClient(address, httpc);
+        Map result = Map.ofEntries(entry("unit", "test"));
+        Map expect = Map.ofEntries(entry("hello", "world"), entry("foo", Map.ofEntries(entry("bar", testIntegerA))));
+
+        try {
+            result = opa.evaluate("policy/echo", Map.ofEntries(
+                entry("hello", "world"),
+                entry("foo", Map.ofEntries(entry("bar", testIntegerA)))
+            ));
+        } catch (OPAException e) {
+            System.out.println("exception: " + e);
+            assertNull(e);
+        }
+
+        assertEquals(expect, result);
+
+        List<LogRecord> recs = handlerObj.getSeenRecs();
+        List<String> logs = new ArrayList<String>();
+        recs.forEach(elem -> logs.add(elem.getLevel().getName() + " " + elem.getMessage()));
+
+        for (String msg : logs) {
+            System.out.printf("DEBUG: log line: %s\n", msg);
+            assertTrue(msg.matches("^INFO LATENCY MEASUREMENT #[0-9]+#/v1/data/policy/echo#$"));
+        }
+    }
+
+    @Test
+    public void testLatencyMeasurementDefaults() {
+        Logger logger = Logger.getLogger("testLatencyMeasurementLogger");
```

Comment:
see this blog for testing loggers: https://dev.to/claudiohigashi/mocking-logger-in-java-with-mockito-51k8

Reply:
So I would have to introduce another dependency in order to do something that already works and is simple? That doesn't seem so good.

Reply:
I ended up DIYing it rather than going with the Mockito approach. It turns out I could just do essentially the same thing I already had using reflection, and the testing needs of this particular feature are trivial enough that it didn't add much complexity.

9a5959928669e504b5945e64e9485a4be013ea19